DEP: Updated `unravel_index()` to support `shape` kwarg

tylerjereddy opened the pull request:
* unravel_index() now supports the `shape` argument
(Fixes #10586) while retaining backwards compatibility
for the `dims` argument

* added corresponding unit tests and docstring
changes

* I note that the warning subclass type that I use might need to
be changed to some type of Deprecation if that's desired?

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1210,7 +1210,25 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
+     * if "dims" is detected in keywords, then replace it with
+     * the new "shape" argument and continue processing as usual */
+
+    PyObject * dims_item;
```
Move this within the `if (kwds)`

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1210,7 +1210,25 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
+        if (dims_item != NULL) {
+            PyErr_WarnEx(PyExc_RuntimeWarning,
+                         "'shape' argument should be used instead of 'dims'",
+                         1);
```
This should be written
```
if (DEPRECATE("the message") < 0) {
    return NULL;
}
```
which correctly deals with warnings being raised as exceptions

eric-wieser commented:
This will need a release note in the compatibility section

The commit prefix should change to `DEP` too, I think.

eric-wieser commented:
Can you rebase on the latest master to fix the circleci error?

tylerjereddy commented:
Revised accordingly; let me know if more refinements are needed.

eric-wieser commented on numpy/lib/tests/test_index_tricks.py:
```diff
@@ -15,6 +16,36 @@
+            assert_equal(np.unravel_index(indices=2,
+                                          dims=(2, 2)),
+                                          (1, 0))
+            assert_(issubclass(w[-1].category, DeprecationWarning))
```
I think there's an `assert_warns` decorator you can just use here

eric-wieser commented on numpy/add_newdocs.py:
```diff
@@ -5386,9 +5386,9 @@ def luf(lamdaexpr, *args, **kwargs):
         this function accepted just one index value.
-    dims : tuple of ints
+    shape : tuple of ints
         The shape of the array to use for unraveling ``indices``.
```
Add:
```
.. versionchanged:: 1.15.0
    Renamed from ``dims`` to ``shape``.

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1210,7 +1210,27 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
+
+
+    if (kwds) {
+        PyObject * dims_item, * shape_item;
```
Elsewhere we write this
```
PyObject *dims_item, *shape_item;
```

eric-wieser commented:
A few tiny additions, but otherwise looks good.

tylerjereddy commented:
I tried to address the second set of revisions -- used a separate commit, but presumably can squash into the original DEP one when things are ready.

mhvk commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1210,7 +1210,27 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
+     * the new "shape" argument and continue processing as usual */
+
+
+    if (kwds) {
```
This is not so bad for a function for which speed is probably not of the essence, but wouldn't you rather look for `dims` only if the parsing failed with the newer order?

tylerjereddy replied:
Is it unambiguously clear that there's a substantially faster alternative implementation that both works and maintains code clarity? Obviously, I'll switch if there is.

The primary reasoning behind my strategy here was to avoid touching the presumably solid keyword Error handling code to the extent possible & just swap the new keyword in at the C API level if the old one is used.

mhvk replied:
It is not a big deal. My main logic was not to hinder the "standard" code path at all.

charris commented:
Need rebase.

tylerjereddy commented:
I rebased / force pushed

eric-wieser commented:
> I tried to address the second set of revisions

Did you accidentally revert this in the rebase?

tylerjereddy commented:
> Did you accidentally revert this in the rebase?

Yes, I will have to fix that.

tylerjereddy commented:
Ok, I *think* I've restored the second set of revisions; also added some more pytest adjustments.

eric-wieser commented:
Looks good to me. Going to leave the pytest things to @charris' judgement - I'm not sure if we're ready to invoke pytest directly yet.

mattip commented on numpy/lib/tests/test_index_tricks.py:
```diff
@@ -14,6 +16,33 @@
+
+        with pytest.raises(TypeError):
+            np.unravel_index(254, ims=(17, 94))
+
```
We should still use the test-framework agnostic suppress_warnings, assert_raises instead of pytest-specific tests

tylerjereddy replied:
Is there a specific reason for that? Still trying to retain compatibility with other frameworks at this point or just want to maintain thematic consistency across the testsuite for now?

mattip replied:
@charris - should we be preferring `assert_raises` or are we ready to commit to pytest?

tylerjereddy commented:
There was some residual resistance to my usage of `pytest.raises()` here so I've switched to the old school `assert_raises` in the hopes of reducing the friction there for now.

Since the PR was stalled on that matter for some time, I've also rebased & tried to bump up the version changed directive and migrate the compatibility note from 1.15 to 1.16.

eric-wieser commented on numpy/lib/tests/test_index_tricks.py:
```diff
@@ -14,6 +16,33 @@
+
+        # test backwards compatibility with older dims
+        # keyword argument; see Issue #10586
+        with pytest.warns(DeprecationWarning):
```
Same comment about `pytest.raises` applies to `pytest.warns`

tylerjereddy commented:
Is there a preference to try removing the `warnings` and `pytest` imports here? Additional requests?

charris commented:
`assert_raises` comes from unittest,  `assert_warns` has some fixes to workaround the python warnings module, and both may be used as context managers. Whether we let NumPy use the pytest versions at some point is not something I feel strongly about, but it is probably best to stay consistent with current usage for the time being.

charris commented:
We try to avoid unneeded imports, pyflakes usually reports them, but we aren't fanatics about it.

tylerjereddy committed:
DEP: Updated unravel_index() kwargs to support `shape`

* unravel_index() now supports the shape argument
(Fixes #10586) while retaining backwards compatibility
for the dims argument

* added corresponding unit tests and docstring
changes

* updated compatibility release notes

tylerjereddy commented:
@charris Thanks for clarifying; I've removed the spurious imports and tests seem to still pass locally--hopefully CI agrees. `pytest` is no longer used in the testing module adjusted here.

tylerjereddy commented:
Should I rebase and resolve merge conflicts or should this just be closed? It looks like I rebased it and shifted the release note to a newer version once already.

mattip commented:
LGTM but needs a rebase to resolve merge conflict.

tylerjereddy commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1156,7 +1156,27 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
+    /* Continue to support the older "dims" argument in place
+     * of the "shape" argument. Issue an appropriate warning
+     * if "dims" is detected in keywords, then replace it with
+     * the new "shape" argument and continue processing as usual */
```
Strictly speaking, this comment is slightly out of line with NumPy formatting standards -- I think this was one of my first C level NumPy PRs.

tylerjereddy replied:
Maybe not a big deal though

tylerjereddy commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1156,7 +1156,27 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
-    char *kwlist[] = {"indices", "dims", "order", NULL};
+    char *kwlist[] = {"indices", "shape", "order", NULL};
+
+    /* Continue to support the older "dims" argument in place
```
```suggestion
    /*
     * Continue to support the older "dims" argument in place
```

tylerjereddy commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1156,7 +1156,27 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
+    /* Continue to support the older "dims" argument in place
+     * of the "shape" argument. Issue an appropriate warning
+     * if "dims" is detected in keywords, then replace it with
+     * the new "shape" argument and continue processing as usual */
```
```suggestion
     * the new "shape" argument and continue processing as usual
     */
```

tylerjereddy commented:
I tried using the new "suggested changes" reviewer interface above to see how that works for this relatively minor adjustment.

tylerjereddy commented:
Looks like either the PR author or assignees can commit the suggestions, but not sure if arbitrary persons with commit rights can.
